# Assignment 2 - Tashlin Reddy
- Donwload the credit dataset 'Credit.csv' from  https://github.com/vahidpartovinia/ycbs255/
- Create the attribute matrix  'X' as a subset of the data that includes only the qualitative attributes 'Income', 'Limit', 'Rating', 'Cards', 'Age', and 'Education'
- Create a discrete response variable, say 'y' by transforming 'Balance' to a binary output, which equals 1 if 'Balance > 1500' and equals 0 otherwise.
- Fit 1- logistic regression, 2- linear discriminant, and 3- quadratic discriminant on the binary Balance as the output variable. Use 'Income', 'Limit', 'Rating', 'Cards', 'Age', and 'Education' as the input variables. 
- Find the probability of (Balance > 1500), for x1 and x2 below, using these three methods. Compare these probabilities and comment.


x1= 'Income' = 63, 'Limit' = 8100, 'Rating' = 600, 'Cards' = 4, 'Age' = 30, 'Education' =14 

x2= 'Income' = 186, 'Limit' = 13414, 'Rating' = 950, 'Cards' = 2, 'Age' = 41, 'Education' =13 

# Submission note
Please fill this jupyter notebook. Extract the pdf file as follows. On  Jupyter manue go to File/Print Preview, then on Browser menu go to File/Print. 

## Only PDF  Submissions will be graded








#### Import dependencies needed for analysis

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
% matplotlib inline

#### Download Credit Dataset

In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/vahidpartovinia/ycbs255/master/jupyter/data/Credit.csv")
df.head()

Unnamed: 0   Income  Limit  Rating  Cards  Age  Education  Gender Student  \
0           1   14.891   3606     283      2   34         11    Male      No   
1           2  106.025   6645     483      3   82         15  Female     Yes   
2           3  104.593   7075     514      4   71         11    Male      No   
3           4  148.924   9504     681      3   36         11  Female      No   
4           5   55.882   4897     357      2   68         16    Male      No   

  Married  Ethnicity  Balance  
0     Yes  Caucasian      333  
1     Yes      Asian      903  
2      No      Asian      580  
3      No      Asian      964  
4     Yes  Caucasian      331

In [3]:
df.describe()

Unnamed: 0      Income         Limit      Rating       Cards  \
count  400.000000  400.000000    400.000000  400.000000  400.000000   
mean   200.500000   45.218885   4735.600000  354.940000    2.957500   
std    115.614301   35.244273   2308.198848  154.724143    1.371275   
min      1.000000   10.354000    855.000000   93.000000    1.000000   
25%    100.750000   21.007250   3088.000000  247.250000    2.000000   
50%    200.500000   33.115500   4622.500000  344.000000    3.000000   
75%    300.250000   57.470750   5872.750000  437.250000    4.000000   
max    400.000000  186.634000  13913.000000  982.000000    9.000000   

              Age   Education      Balance  
count  400.000000  400.000000   400.000000  
mean    55.667500   13.450000   520.015000  
std     17.249807    3.125207   459.758877  
min     23.000000    5.000000     0.000000  
25%     41.750000   11.000000    68.750000  
50%     56.000000   14.000000   459.500000  
75%     70.000000   16.000000   863.000000  
max     98.000000   20.000000  1999.000000

In [4]:
df.shape

(400, 12)

#### Create the attribute matrix 'X' as a subset of the data that includes only the qualitative attributes 'Income', 'Limit', 'Rating', 'Cards', 'Age', and 'Education'

In [5]:
X = df[['Income','Limit','Rating','Cards','Age','Education']].values #convert to numpy array for efficiency

#### Create a discrete response variable, say 'y' by transforming 'Balance' to a binary output, which equals 1 if 'Balance > 1500' and equals 0 otherwise.

In [6]:
y = df[['Balance']]

In [7]:
y.describe()

Balance
count   400.000000
mean    520.015000
std     459.758877
min       0.000000
25%      68.750000
50%     459.500000
75%     863.000000
max    1999.000000

In [8]:
y['Balance'] = np.where(y['Balance'] > 1500, 1, 0)

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [9]:
y.describe()

Balance
count  400.000000
mean     0.022500
std      0.148489
min      0.000000
25%      0.000000
50%      0.000000
75%      0.000000
max      1.000000

#### Double checking to see if our change to Balance values makes sense, which it does

In [10]:
y = y['Balance'].values #convert to numpy array again

In [11]:
#import Machine Learning libraries
from sklearn.linear_model import LogisticRegression
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis, QuadraticDiscriminantAnalysis

#### Fit 1- logistic regression, 2- linear discriminant, and 3- quadratic discriminant on the binary Balance as the output variable. Use 'Income', 'Limit', 'Rating', 'Cards', 'Age', and 'Education' as the input variables.

In [12]:
lr = LogisticRegression()
lr.fit(X,y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [13]:
lda = LinearDiscriminantAnalysis()
lda.fit(X,y)

LinearDiscriminantAnalysis(n_components=None, priors=None, shrinkage=None,
              solver='svd', store_covariance=False, tol=0.0001)

In [14]:
qda = QuadraticDiscriminantAnalysis()
qda.fit(X,y)

QuadraticDiscriminantAnalysis(priors=None, reg_param=0.0,
               store_covariance=False, store_covariances=None, tol=0.0001)

In [15]:
print(lr.intercept_, lr.coef_)

[-0.81425465] [[-0.0095085   0.00499375 -0.05897041  0.11844931 -0.05977197 -0.35403372]]


###  Find the probability of (Balance > 1500), for x1 and x2 below, using these three methods. Compare these probabilities and comment.
#### x1= 'Income' = 63, 'Limit' = 8100, 'Rating' = 600, 'Cards' = 4, 'Age' = 30, 'Education' =14

#### x2= 'Income' = 186, 'Limit' = 13414, 'Rating' = 950, 'Cards' = 2, 'Age' = 41, 'Education' =13

In [16]:
X_pred = np.array([63, 8100, 600, 4, 30, 14, 186, 13414, 950, 2, 41, 13]).reshape(2,-1)


In [17]:
print(lr.predict_proba(X_pred))

[[0.93226322 0.06773678]
 [0.17289292 0.82710708]]


In [18]:
print(lda.predict_proba(X_pred))

[[0.94144572 0.05855428]
 [0.00721199 0.99278801]]


In [19]:
print(qda.predict_proba(X_pred))

[[9.99999999e-01 1.24419207e-09]
 [7.83057752e-04 9.99216942e-01]]


#### Comments: 

Predictions:

Logistic Regression predicts a 6.77% chance x1 and a 82.7% chance x2 have balances greater than 1500. 

Linear Discriminant Analysis predicts a 5.86% chance x1 and a 99.3% chance x2 have balances greater than 1500.

Quadratic Discriminant Analysis predicst a ~0% chance x1 and a ~99.99% chance x2 have balances greater than 1500.


At first glance it would seem QDA is a bit extreme. It might not be flexible enough to handle new predictions.

LDA seems a bit better, but I still think it's a bit extreme predicting x2. Ideally it would be great if we could say would that we are 99.3% certain this person(x2) would have a balance greater than 1500, but we'd have to do a deeper dive into the analysis to be more confident. 

Logistic Regression seems to be the best but again, it's just an assumption at this point, without more analysis.

#### Cross Validation should give us better insight into the data

In [21]:
from sklearn.model_selection import KFold
k = 10
rss_lr = np.zeros(k)
rss_lda = np.zeros(k)
rss_qda = np.zeros(k)

kf = KFold(n_splits=k, shuffle=True)
i = 0
for train_i, test_i in kf.split(df):
    lr = LogisticRegression()
    lda = LinearDiscriminantAnalysis()
    qda = QuadraticDiscriminantAnalysis()
    lr = lr.fit(X[train_i], y[train_i])
    qda = qda.fit(X[train_i], y[train_i])
    lda = lda.fit(X[train_i], y[train_i])
    
    
    rss_lr[i]=np.sum((lr.predict(X[test_i]) - y[test_i])**2)
    rss_lda[i]=np.sum((lda.predict(X[test_i]) - y[test_i])**2)
    rss_qda[i]=np.sum((qda.predict(X[test_i]) - y[test_i])**2)
    i+=1

print("logistic reg: rss =", np.sum(rss_lr))
print("lda: rss =", np.sum(rss_lda))
print("qda: rss =", np.sum(rss_qda))


logistic reg: rss = 8.0
lda: rss = 13.0
qda: rss = 11.0


#### Summing up the Residual Sum of Squares, cross validation verifies our assumption that Logistic Regression seems to be the best predictor in this case

#### Surprisingly QDA seems to be a better predictor than LDA. 

#### Overall, from the analysis done here, Logistic Regression is the best predictor from the 3.